pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -74,21 +76,39 @@ func GetDefaultCircuitBreakerThresholds(direction model.TrafficDirection) *v2Clu
 // For inbound (sidecar only): Cluster for each inbound endpoint port and for each service port
 func (configgen *ConfigGeneratorImpl) BuildClusters(env *model.Environment, proxy *model.Proxy, push *model.PushContext) ([]*apiv2.Cluster, error) {
 	clusters := make([]*apiv2.Cluster, 0)
+	instances := proxy.ServiceInstances
+
+	// compute the proxy's locality. See if we have a CDS cache for that locality.
+	// If not, compute one.
+	locality := proxy.Locality
+	if locality == nil {
+		// Get the locality from the proxy's service instances.
+		// We expect all instances to have the same locality. So its enough to look at the first instance
+		if len(instances) > 0 {
+			locality = util.ConvertLocality(instances[0].GetLocality())
```

Comment:
Confused - how will it work for 'batch' jobs or clients ( who do not expose any Service) ? 
And why do you get locality from ServiceInstance in the first place - node should pass the locality as metadata already.

Reply:
If node doesn't know its own locality - we may need to lookup the pod by IP .

ServiceInstance is only populated if there is a Service associated with the pod, which is not the case (and is a very flaky/complicated logic, depending on readiness and lots of other things).

Reply:
See the line 83 above.. the locality is looked up from the service instances if and only if the node does not supply its own locality. With regard to batch jobs, this code is not doing anything different. CDS code is using the proxy service instances for inbound clusters. So, if there are service instances for batch jobs (which I bet there aren't any), then we lookup locality from them. But the most likely case is that batch jobs dont have service instances - this loop wont even execute